File: apollo-runtime/src/main/java/com/apollographql/apollo/cache/CacheHeaders.java
package com.apollographql.apollo.cache;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

import javax.annotation.Nullable;

/**
 * A key/value collection which is sent with {@link com.apollographql.apollo.cache.normalized.Record}
 * from a {@link com.apollographql.apollo.api.Operation} to the
 * {@link com.apollographql.apollo.cache.normalized.NormalizedCache}.
 *
 * For headers which the default {@link com.apollographql.apollo.cache.normalized.NormalizedCache} respect, see
 * {@link ApolloCacheHeaders}.
 */
public final class CacheHeaders {

  private final Map<String, String> headerMap;

  public static CacheHeaders.Builder builder() {
    return new Builder();
  }

  public static final CacheHeaders NONE = new CacheHeaders(Collections.<String, String>emptyMap());

  public static final class Builder {

    private final Map<String, String> headerMap = new LinkedHashMap<>();

    public Builder addHeader(String headerName, String headerValue) {
      headerMap.put(headerName, headerValue);
      return this;
    }

    public Builder addHeaders(Map<String, String> headerMap) {
      this.headerMap.putAll(headerMap);
      return this;
    }

    public CacheHeaders build() {
      return new CacheHeaders(headerMap);
    }
  }

  /**
   * @return A {@link CacheHeaders.Builder} with a copy of this {@link CacheHeaders} values.
   */
  public CacheHeaders.Builder toBuilder() {
    CacheHeaders.Builder builder = builder();
    builder.addHeaders(headerMap);
    return builder;
  }

  CacheHeaders(Map<String, String> headerMap) {
    this.headerMap = headerMap;
  }

  @Nullable
  public String headerValue(String header) {
    return headerMap.get(header);
  }

  public boolean hasHeader(String headerName) {
    return headerMap.containsKey(headerName);
  }
}

File: libraries/apollo-normalized-cache-api-incubating/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/api/CacheHeaders.kt
package com.apollographql.apollo3.cache.normalized.api

import com.apollographql.apollo3.annotations.ApolloDeprecatedSince
import com.apollographql.apollo3.api.Operation
import kotlin.jvm.JvmField
import kotlin.jvm.JvmStatic

/**
 * A key/value collection which is sent with [Record] from a [Operation] to the [NormalizedCache].
 *
 * For headers which the default [NormalizedCache] respect, see [ApolloCacheHeaders].
 */
class CacheHeaders internal constructor(private val headerMap: Map<String, String>) {

  class Builder {
    private val headerMap = mutableMapOf<String, String>()

    fun addHeader(headerName: String, headerValue: String) = apply {
      headerMap[headerName] = headerValue
    }

    fun addHeaders(headerMap: Map<String, String>) = apply {
      this.headerMap.putAll(headerMap)
    }

    fun build() = CacheHeaders(headerMap)
  }

  /**
   * @return A [CacheHeaders.Builder] with a copy of this [CacheHeaders] values.
   */
  @Deprecated("Use newBuilder() instead", ReplaceWith("newBuilder()"))
  @ApolloDeprecatedSince(ApolloDeprecatedSince.Version.v3_3_1)
  fun toBuilder(): Builder = newBuilder()

  fun newBuilder(): Builder = builder().addHeaders(headerMap)

  fun headerValue(header: String): String? = headerMap[header]

  fun hasHeader(headerName: String): Boolean = headerMap.containsKey(headerName)

  companion object {
    @JvmStatic
    fun builder() = Builder()

    @JvmField
    val NONE = CacheHeaders(emptyMap())
  }
  operator fun plus(cacheHeaders: CacheHeaders): CacheHeaders {
    return newBuilder().addHeaders(cacheHeaders.headerMap).build()
  }
}
